### PR TITLE
Refactor: convert DependenciesInfoTask to Java

### PR DESCRIPTION
**Pull request by dRadest**
Contributes to #34459 

**status**: ongoing

Done so far:
* refactored DependenciesInfoTask.groovy to DependenciesInfoTask.java

### CODE REVIEW

File: buildSrc/src/main/java/org/elasticsearch/gradle/DependenciesInfoTask.java [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+     */
+    protected String getLicenseType(final String group, final String name) throws IOException {
+        File license;
+
+        if (licensesDir.exists()) {
+            File[] matchedfiles = licensesDir.listFiles((dir, name) -> name.contains("-LICENSE"));
+            for(File file : matchedfiles) {
+                String prefix = file.getName().split("-LICENSE.*")[0];
+                if (group.contains(prefix) || name.contains(prefix)) {
+                    license = file.getAbsoluteFile();
```
Comment (rhamedy) [inline buildSrc/src/main/java/org/elasticsearch/gradle/DependenciesInfoTask.java]: I know this is same as existing code but, makes me curious. 

I wonder how this `loop` works? When the `if (group.contains(prefix) || name.contains(prefix))` is true more than `once` then previous value of `license` is replaced with `new one` 🤔If it is `true` only `once` then we should `return` early after assigning `license` value. No? 😕

File: buildSrc/src/main/java/org/elasticsearch/gradle/DependenciesInfoTask.java [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+                }
+            }
+        }
+
+        if (license != null) {
+            // replace * because they are sometimes used at the beginning lines as if the license was a multi-line comment
+            final String content = new String(Files.readAllBytes(outfile.toPath()), StandardCharsets.UTF_8);
+            content.replaceAll("\\s+", " ").replaceAll("\\*", " ");
```
Comment (rhamedy) [inline buildSrc/src/main/java/org/elasticsearch/gradle/DependenciesInfoTask.java]: It kinda feels weird when invoking `replaceAll` on a `final` instance. Also `replaceAll` does not change the value of `content` and it returns the `updated` string and you are not storing it anywhere (hence changes lost). You might have to stick with existing solution 

```final String content = new String(license.readBytes(), "UTF-8").replaceAll("\\s+", " ").replaceAll("\\*", " ")``` 

It does not change the value of a `final` field as well as the result of `replaceAll` on string is not `lost` 👍

**Review by rhamedy — COMMENTED** [review]
Not part of the `Elastic` team but, left some comments anyway based on my learning experience from my PR for `39605` 👍

File: buildSrc/src/main/java/org/elasticsearch/gradle/DependenciesInfoTask.java [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.gradle;
+
+import org.gradle.api.DefaultTask;
+import org.elasticsearch.gradle.precommit.DependencyLicensesTask;
```
[the remark this answers is not in the document]
Comment (dRadest) [reply buildSrc/src/main/java/org/elasticsearch/gradle/DependenciesInfoTask.java]: Yep, that was my general idea, i.e. to wait

File: buildSrc/src/main/java/org/elasticsearch/gradle/DependenciesInfoTask.java [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+    private Configuration compileOnlyConfiguration;
+    
+    private LinkedHashMap<String, String> mappings;
+
+    /** Directory to read license files */
+    private File licensesDir = new File(getProject().getProjectDir(), "licenses");
+
+    private File outputFile = new File(getProject().getBuildDir(), "reports/dependencies/dependencies.csv");
+
+    @Input
```
[the remark this answers is not in the document]
Comment (dRadest) [reply buildSrc/src/main/java/org/elasticsearch/gradle/DependenciesInfoTask.java]: Thanks. I missed that. Is this just for the ```@input``` or all annotations?

File: buildSrc/src/main/java/org/elasticsearch/gradle/DependenciesInfoTask.java (hunk 1 above)
Comment (dRadest) [reply buildSrc/src/main/java/org/elasticsearch/gradle/DependenciesInfoTask.java]: I'm not sure, but I wonder if there could be more than one license file for a dependency? How would we know without looping through all of them? And if there are, looping through all of them and sticking to the last one makes as much sense as any other.

File: buildSrc/src/main/java/org/elasticsearch/gradle/DependenciesInfoTask.java (hunk 2 above)
Comment (dRadest) [reply buildSrc/src/main/java/org/elasticsearch/gradle/DependenciesInfoTask.java]: Ups! My bad. Thanks for pointing that out. But isn't ```readBytes()``` [groovy](http://docs.groovy-lang.org/latest/html/groovy-jdk/java/io/File.html#readBytes()) ? I replaced it with ```Files.readAllBytes()```

**Comment by atorok** [comment]
Thanks for picking this up @dRadest , and thanks for pitching in @rhamedy ! 
@elasticmachine  test this please.

File: buildSrc/src/main/java/org/elasticsearch/gradle/DependenciesInfoTask.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Megamiun) [reply buildSrc/src/main/java/org/elasticsearch/gradle/DependenciesInfoTask.java]: Sorry, I had some problems running the tests and stopped working on #35231. I'll try to see if I can make the tests work this weekend.

**Comment by atorok** [comment]
@elasticmachine update branch

**Comment by atorok** [comment]
@elasticmachine test this please

**Comment by atorok** [comment]
@dRadest looks like something is not wired correctly ? CI is picking up on ```15:16:29 > Task :buildSrc:compileJava
15:16:29 /var/lib/jenkins/workspace/elastic+elasticsearch+pull-request-1/buildSrc/src/main/java/org/elasticsearch/gradle/DependenciesInfoTask.java:30: error: cannot find symbol
15:16:29 import org.elasticsearch.gradle.precommit.DependencyLicensesTask;
15:16:29                                          ^
15:16:29   symbol:   class DependencyLicensesTask
15:16:29   location: package org.elasticsearch.gradle.precommit```

**Comment by atorok** [comment]
I'm going to close this. @dRadest feel free to open it if you plan to work on it again

**Pull request closed without merging**